test/scope/scope.test.ts

```diff
@@ -0,0 +1,298 @@
+
+      class DynamicValue {
+        constructor(readonly inRootRequestScope:InRootRequestScope){
+
+        }
+      }
+
+      class FromFactory {
+        constructor(readonly inRootRequestScope:InRootRequestScope){
+
+        }
+      }
+
+      @injectable()
+      class Root {
+        constructor(
+          @inject(TYPES.rootRequestScope) readonly inRootRequestScope:InRootRequestScope,
+          @inject(TYPES.factory) readonly factory:()=>FromFactory,
+          @inject(TYPES.dynamicValue) readonly dynamicValue:DynamicValue
+          ){
+          }
+
+      }
+      const trueFalse = [true,false];
+      trueFalse.forEach(isRootRequestScope => {
+        const container = new Container();
+        const bindIn = container.bind<InRootRequestScope>(TYPES.rootRequestScope).to(InRootRequestScope);
+        if(isRootRequestScope){
+          bindIn.inRootRequestScope();
+        }else{
+          bindIn.inRequestScope();
+        }
+
+        container.bind(TYPES.factory).toFactory(context => {
+          const inRootRequestScope = context.container.get<InRootRequestScope>(TYPES.rootRequestScope);
+          return () => {
+            return new FromFactory(inRootRequestScope);
+          }
+        })
+        container.bind(TYPES.dynamicValue).toDynamicValue(context =>
+          new DynamicValue(context.container.get<InRootRequestScope>(TYPES.rootRequestScope)));
+        const root = container.resolve(Root);
+
+        expect(root.inRootRequestScope === root.factory().inRootRequestScope).to.equal(isRootRequestScope);
+        expect(root.inRootRequestScope === root.dynamicValue.inRootRequestScope).to.equal(isRootRequestScope);
+      })
+
+    });
+
+    it("should be possible to delay by pushing the context to the context stack", () => {
+      const TYPES = {
+        rootRequestScope:'rootRequestScope',
+        factory:'factory',
+      }
+
+      @injectable()
+      class InRootRequestScope{
+
+      }
+
+      class FromFactory {
+        constructor(readonly inRootRequestScope:InRootRequestScope){
+
+        }
+      }
+
+      @injectable()
+      class Root {
+        constructor(
+          @inject(TYPES.rootRequestScope) @named("one") readonly inRootRequestScope:InRootRequestScope,
+          @inject(TYPES.factory) readonly factory:(name:"one"|"theother")=>FromFactory,
+          ){
+          }
+
+      }
+      const trueFalse = [true,false];
+      trueFalse.forEach(isRootRequestScope => {
+        const container = new Container();
+        const bindTo = container.bind<InRootRequestScope>(TYPES.rootRequestScope).to(InRootRequestScope);
+        if(isRootRequestScope){
+          bindTo.inRootRequestScope();
+        }else{
+          bindTo.inRequestScope();
+        }
+
+        container.bind(TYPES.factory).toFactory(context => {
+          return (name:"one"|"theother") => {
+            const inRootRequestScope = context.inRootRequestScope().container.getNamed<InRootRequestScope>(TYPES.rootRequestScope,name)
```

Comment:
This method of retaining the context hierarchy does not work for toAutoFactory.  Should we replicate ?
Current code
```

    public toAutoFactory<T2>(serviceIdentifier: interfaces.ServiceIdentifier<T2>): interfaces.BindingWhenOnSyntax<T> {
        return this.toFactory((context) => {
            const autofactory = () => context.container.get<T2>(serviceIdentifier);
            return autofactory;
        });
    }
```